proposals/patterns3.md

```diff
@@ -0,0 +1,253 @@
+- https://github.com/dotnet/csharplang/issues/1350 Negated `not` patterns that require a given pattern *not* to match; and
+- https://github.com/dotnet/csharplang/issues/812 Relational patterns that require the input value to be less than, less than or equal to, etc a given constant.
+
+## Parenthesized Patterns
+
+Parenthesized patterns permit the programmer to put parentheses around any pattern.  This is not so useful with the existing patterns in C# 8.0, however the new pattern combinators introduce a precedence that the programmer may want to override.
+
+```antlr
+primary_pattern
+    : parenthesized_pattern
+    ;
+parenthesized_pattern
+    : '(' pattern ')'
+    ;
+```
+
+## Relational Patterns
+
+Relational patterns permit the programmer to express that an input value must satisfy a relational constraint when compared to a constant value:
+
+``` C#
+    public static LifeStage LifeStageAtAge(int age) => age switch
+    {
+        < 0 =>  LiftStage.Prenatal,
+        < 2 =>  LifeStage.Infant,
+        < 4 =>  LifeStage.Toddler,
+        < 6 =>  LifeStage.EarlyChild,
+        < 12 => LifeStage.MiddleChild,
+        < 20 => LifeStage.Adolescent,
+        < 40 => LifeStage.EarlyAdult,
+        < 65 => LifeStage.MiddleAdult,
+        _ =>    LifeStage.LateAdult,
+    };
+```
+
+We imagine supporting `<`, `<=`, `>`, and `>=` patterns on all of the built-in types that support such binary relational operators with two operands of the same type in an expression. Specifically, we support all of these relational patterns for `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, and `decimal`.
+
+```antlr
+primary_pattern
+    : relational_pattern
+    ;
+relational_pattern
+    : '<' relational_expression
+    | '<=' relational_expression
+    | '>' relational_expression
+    | '>=' relational_expression
+    ;
+```
+
+The expression is required to evaluate to a constant value.  It is an error if that constant value is `double.NaN` or `float.NaN`.  It is an error if the expression is a null constant and the relational operator is `<`, `<=`, `>`, or `>=`.
+
+When the input is a type for which a suitable built-in binary relational operator is defined that is applicable with the input as its left operand and the given constant as its right operand, the evaluation of that operator is taken as the meaning of the relational pattern.  Otherwise we convert the input to the type of the expression using an explicit nullable or unboxing conversion.  It is a compile-time error if no such conversion exists.  The pattern is considered not to match if the conversion fails.  If the conversion succeeds then the result of the pattern-matching operation is the result of evaluating the expression `e OP v` where `e` is the converted input, `OP` is the relational operator, and `v` is the constant expression.
```

Comment:
> explicit nullable or unboxing conversion

What if there's an implicit user-defined conversion to the appropriate type? For example, if you have a custom fractional struct, you could define a conversion operator to decimal, and it would be nice if you could then pattern match on that type. Given that implicit conversions are supposed to never lose information, it seems like this should be fine.

Reply:
Actually, after considering for a bit, this is probably a very bad thing that would lead to inconsistent behavior.